lib/jsonrpc_client.ml

```diff
@@ -17,49 +17,101 @@
 module D = Debug.Make(struct let name = "jsonrpc_client" end)
 open D
 
-let input_json_object fin =
-	let buf = Buffer.create 1024 in
-	let brace_cnt = ref 0 in
-	let in_string = ref false in
-	let last_char () = Buffer.nth buf (Buffer.length buf - 1) in
-	let rec get () =
-		let c = input_char fin in
+exception Timeout
+
+let json_rpc_max_len = ref 65536 (* Arbitrary maximum length of RPC response *)
+let json_rpc_read_timeout = ref "60000000000"
+let json_rpc_write_timeout = ref "60000000000"
+
+let to_s  s = (Int64.to_float s) *. 1e-9
+
+(* Read the entire contents of the fd, of unknown length *)
+let timeout_read fd timeout =
+	let buf = Buffer.create !json_rpc_max_len in
+	let rec inner max_time max_bytes =
+		let c = Mtime_clock.counter () in
+		let get_used_time counter = Mtime.Span.to_uint64_ns (Mtime_clock.count counter) in
+		let (ready_to_read, _, _) = Unix.select [fd] [] [] (to_s max_time) in
+		if List.mem fd ready_to_read
+		then
 		begin
-			match c with
-			| '{' when not !in_string -> brace_cnt := !brace_cnt + 1
-			| '}' when not !in_string -> brace_cnt := !brace_cnt - 1
-			| '"' when !in_string && (last_char () <> '\\') -> in_string := false
-			| '"' when not !in_string -> in_string := true
-			| _ -> ()
-		end;
-		Buffer.add_char buf c;
-		if !brace_cnt > 0 then
-			get ()
+			let bytes = Bytes.make 4096 '\000' in
+			match Unix.read fd bytes 0 4096 with
+				| 0 -> Buffer.contents buf (* EOF *)
+				| n ->
+					if n > max_bytes
+					then
+					begin
+						debug "exceeding maximum read limit %d, clear buffer" !json_rpc_max_len;
+						Buffer.clear buf; (* otherwise might be parse error *)
+						Buffer.contents buf
+					end
+					else
+					begin
+						let used_time = get_used_time c in
+						let remain_time = Int64.sub max_time used_time in
```

Comment:
For sake of simplicity, the time calculation in select-and-return-ready branch could be removed. The worst case is for each select-and-return-ready, it spent `max_time - e'` time (`e'` is a small time), and it read only a few bytes from the ready. I think this is quite rare case.

Reply:
I personally prefer a more accurate calculation, than just a few lines's change/simplicity.